include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,237 @@
+	const bool run = !skip;
+
+	if (run && is_bxdf_or_combiner)
 	{
-#ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
-#endif
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
+	}
 
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
 
+	if (run && is_bxdf && positiveCosFactor)
+	{
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
 
-#ifndef ONLY_ONE_NDF
-		} else
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		bool is_valid = (NdotV > FLT_MIN);
+		bool refraction = false;
+#ifdef OP_DIELECTRIC
+		if (op == OP_DIELECTRIC && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+		{
+			is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, eta, rcp_eta);
+			refraction = true;
+		}
+		else
+#endif
+		{
+			microfacet = _microfacet;
+		}
+
+#if defined(OP_DIELECTRIC) || defined(OP_CONDUCTOR)
+		is_valid = irr_glsl_isValidVNDFMicrofacet(microfacet.isotropic, is_bsdf, refraction, s.VdotL, eta, rcp_eta);
+#endif
+		const vec3 albedo = params_getReflectance(params);
+		const float a = params_getAlpha(params);
+		const float a2 = a * a;
+
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
+		{
+			bxdf_eval = albedo * irr_glsl_oren_nayar_cos_eval(s, currInteraction.isotropic, a2);
+		}
+		else
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+		if (is_valid && a > ALPHA_EPSILON)
+		{
+			float bxdf_eval_scalar_part;
+			const uint ndf = instr_getNDF(instr);
+
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float ay = params_getAlphaV(params);
+			const float ay2 = ay*ay;
 #endif
+			const float NdotL = cosFactor;
+			const float NdotL2 = s.NdotL2;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotL2 = s.TdotL * s.TdotL;
+			const float BdotL2 = s.BdotL * s.BdotL;
 #endif
 
-#ifdef NDF_BECKMANN
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_BECKMANN) {
+			const float NdotV2 = currInteraction.isotropic.NdotV_squared;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+			const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
 #endif
 
+			const float NdotH = microfacet.isotropic.NdotH;
+			const float NdotH2 = microfacet.isotropic.NdotH2;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotH2 = microfacet.TdotH * microfacet.TdotH;
+			const float BdotH2 = microfacet.BdotH * microfacet.BdotH;
+#endif
```

Comment:
why don't we precompute the squared microfacet and interaction stuff?
(obvs just for this backend, not in the general engine-wide structs)

Reply:
if we dont in general, why would we here?

Reply:
because here it would be useful I guess?

Reply:
I'm not suggesting adding fields to `irr_glsl_AnisotropicViewSurfaceInteraction ` or `irr_glsl_AnisotropicMicrofacetCache`

just creating 1 struct each, for the usage of the material compiler glsl backend which declares a the above as a member, plus the extra fields (poor man's inheritance)